sdk/python/starwhale/data_store.py

```diff
@@ -0,0 +1,703 @@
+        self.uri = uri
+        self.display_text = display_text
+        self.mime_type = mime_type
+
+    def __str__(self) -> str:
+        return json.dumps({"uri": self.uri, "display_text": self.display_text, "mime_type": self.mime_type})
+
+    def __eq__(self, other) -> bool:
+        return (
+            isinstance(other, Link)
+            and self.uri == other.uri
+            and self.display_text == other.display_text
+            and self.mime_type == other.mime_type
+        )
+
+
+NONE = Type("none", None, 1, None)
+INT8 = Type("int", pa.int8(), 8, 0)
+INT16 = Type("int", pa.int16(), 16, 0)
+INT32 = Type("int", pa.int32(), 32, 0)
+INT64 = Type("int", pa.int64(), 64, 0)
+FLOAT16 = Type("float", pa.float16(), 16, 0.0)
+FLOAT32 = Type("float", pa.float32(), 32, 0.0)
+FLOAT64 = Type("float", pa.float64(), 64, 0.0)
+BOOL = Type("bool", pa.bool_(), 1, 0)
+STRING = Type("str", pa.string(), 32, "")
+BYTES = Type("bytes", pa.binary(), 32, b"")
+LINK = LinkType()
+
+_TYPE_DICT: dict[Type, Type] = {
+    type(None): NONE,
+    np.byte: INT8,
+    np.int8: INT8,
+    np.int16: INT16,
+    np.int32: INT32,
+    np.int64: INT64,
+    int: INT64,
+    np.float16: FLOAT16,
+    np.float32: FLOAT32,
+    np.float_: FLOAT64,
+    float: FLOAT64,
+    np.bool_: BOOL,
+    bool: BOOL,
+    str: STRING,
+    bytes: BYTES,
+    Link: LINK,
+}
+
+_TYPE_NAME_DICT = {str(v): v for k, v in _TYPE_DICT.items()}
+
+
+def _get_type(obj: Any) -> Type:
+    return _TYPE_DICT.get(type(obj), None)
+
+
+class ColumnSchema:
+    def __init__(self, name: str, type: Type):
+        self.name = name
+        self.type = type
+
+    def __eq__(self, other) -> bool:
+        return isinstance(other, ColumnSchema) and self.name == other.name and self.type is other.type
+
+
+class TableSchema:
+    def __init__(self, key_column: str, columns: list[ColumnSchema]):
+        self.key_column = key_column
+        self.columns = {col.name: col for col in columns}
+
+    def copy(self) -> "TableSchema":
+        return TableSchema(self.key_column, self.columns.values())
+
+    @staticmethod
+    def parse(json_str: str) -> "TableSchema":
+        d = json.loads(json_str)
+        return TableSchema(d["key"], [ColumnSchema(col["name"], _TYPE_NAME_DICT[col["type"]]) for col in d["columns"]])
+
+    def __str__(self) -> str:
+        return json.dumps(
+            {
+                "key": self.key_column,
+                "columns": [{"name": col.name, "type": str(col.type)} for col in self.columns.values()],
+            }
+        )
+
+    def __repr__(self) -> str:
+        return self.__str__()
+
+    def __eq__(self, other) -> bool:
+        return isinstance(other, TableSchema) and self.key_column == other.key_column and self.columns == other.columns
+
+
+def _get_table_path(root_path: str, table_name: str) -> str:
+    return os.path.join(root_path, *table_name.split("/"))
```

Comment:
if the `table_name` is a path split string, `os.path.join` can handle it automatically.

```
In [1]: import os

In [2]: os.path.join("a", "b/c/d/e")
Out[2]: 'a/b/c/d/e'
```

Reply:
it does not work on Windows, which will be "a\\b/c/d/e"

Reply:
The `pathlib` is maybe  the better choice.

- in linux

```python
In [1]: import platform
In [2]: platform.system()
Out[2]: 'Linux'
In [3]: from pathlib import Path
In [4]: Path("/tmp") / "b/c/d"
Out[4]: PosixPath('/tmp/b/c/d')
```

- in windows

```python
>>> import platform
>>> platform.system()
'Windows'
>>> from pathlib import Path
>>> Path("/tmp") / "b/c/d"
WindowsPath('/tmp/b/c/d')
>>> Path("/tmp") / "b\c\d"
WindowsPath('/tmp/b/c/d')
```
cc @xuchuan

Reply:
done